src/synth/fluid_synth.c

```diff
@@ -5600,190 +5753,305 @@ fluid_synth_set_chorus_on(fluid_synth_t *synth, int on)
 }
 
 /**
  * Set the chorus type. See fluid_synth_set_chorus() for further info.
  * @return #FLUID_OK on success, #FLUID_FAILED otherwise
+ * @deprecated Use fluid_synth_chorus_set_param instead.
  */
 int fluid_synth_set_chorus_type(fluid_synth_t *synth, int type)
 {
-    return fluid_synth_set_chorus_full(synth, FLUID_CHORUS_SET_TYPE, 0, 0, 0, 0, type);
+    return fluid_synth_chorus_set_param(synth, -1, FLUID_CHORUS_TYPE, type);
 }
 
+/**
+ * Set one chorus parameter to one fx group.
+ * @param synth FluidSynth instance.
+ * @param fx_group index of the fx group to which parameters must be set.
+ *  must be in the range [-1..synth->effects_groups[. If -1 the changes are
+ *  applied to all fx groups.
+ * @param enum indicating the parameter to set (#fluid_chorus_param).
+ *  FLUID_CHORUS_NR, chorus voice count (0-99, CPU time consumption proportional to
+ *  this value).
+ *  FLUID_CHORUS_LEVEL, chorus level (0.0-10.0).
+ *  FLUID_CHORUS_SPEED, chorus speed in Hz (0.1-5.0).
+ *  FLUID_CHORUS_DEPTH, chorus depth (max value depends on synth sample-rate,
+ *   0.0-21.0 is safe for sample-rate values up to 96KHz).
+ *  FLUID_CHORUS_TYPE, chorus waveform type (#fluid_chorus_mod)
+ * @value, parameter value
+ * @return #FLUID_OK on success, #FLUID_FAILED otherwise.
+ */
 int
-fluid_synth_set_chorus_full(fluid_synth_t *synth, int set, int nr, double level,
-                            double speed, double depth_ms, int type)
+fluid_synth_chorus_set_param(fluid_synth_t *synth, int fx_group, enum fluid_chorus_param param,
+                             double value)
 {
     int ret;
-    fluid_rvoice_param_t param[MAX_EVENT_PARAMS];
+    double values[FLUID_CHORUS_PARAM_LAST] = {0.0};
 
-    fluid_return_val_if_fail(synth != NULL, FLUID_FAILED);
-    /* if non of the flags is set, fail */
-    fluid_return_val_if_fail(set & FLUID_CHORUS_SET_ALL, FLUID_FAILED);
+    /* setting name (except lfo waveform type) */
+    static const char *name[FLUID_CHORUS_PARAM_LAST-1] =
+    {
+        "synth.chorus.nr", "synth.chorus.level",
+        "synth.chorus.speed", "synth.chorus.depth"
+    };
 
-    /* Synth shadow values are set here so that they will be returned if queried */
+    double f_min; /* minimum value */
+    double f_max; /* maximum value */
+
+    /* check parameters */
+    fluid_return_val_if_fail(synth != NULL, FLUID_FAILED);
+    fluid_return_val_if_fail((param >= 0) && (param < FLUID_CHORUS_PARAM_LAST), FLUID_FAILED);
     fluid_synth_api_enter(synth);
 
-    if(set & FLUID_CHORUS_SET_NR)
+    if(fx_group  < -1 || fx_group >= synth->effects_groups)
     {
-        synth->chorus_nr = nr;
+        FLUID_API_RETURN(FLUID_FAILED);
     }
 
-    if(set & FLUID_CHORUS_SET_LEVEL)
+    /* check if chorus value is in max min range */
+    if(param == FLUID_CHORUS_TYPE) /* integer*/
     {
-        synth->chorus_level = level;
+        f_min = (double)FLUID_CHORUS_MOD_SINE;
+        f_max = (double)FLUID_CHORUS_MOD_TRIANGLE;
```

Comment:
Casting an enum to double really strikes me. Counter proposal: Instead of simply passing `int on` to `fluid_synth_chorus_on()`, we provide it with the `enum fluid_chorus_mod`. Additionally, we'll need an extra value that defines `FLUID_CHORUS_OFF`.